components/file_system/src/iosnoop/biosnoop.rs

```diff
@@ -0,0 +1,508 @@
+                .unwrap();
+            let mut r = vec![A512::default(); 2];
+            assert_ne!(f.read(&mut r.as_bytes_mut()).unwrap(), 0);
+            drop(f);
+        })
+        .join()
+        .unwrap();
+
+        let delta = ctx.delta_and_refresh();
+        assert_eq!(delta.get(&IOType::Compaction).unwrap().write, 0);
+        assert_eq!(delta.get(&IOType::Compaction).unwrap().read, 0);
+        assert_eq!(delta.get(&IOType::Other).unwrap().write, 0);
+        assert_ne!(delta.get(&IOType::Other).unwrap().read, 0);
+
+        flush_io_metrics();
+        assert_ne!(IO_LATENCY_MICROS_VEC.compaction.write.get_sample_count(), 0);
+        assert_ne!(IO_LATENCY_MICROS_VEC.other.read.get_sample_count(), 0);
+        assert_ne!(IO_BYTES_VEC.compaction.write.get(), 0);
+        assert_ne!(IO_BYTES_VEC.other.read.get(), 0);
+    }
+
+    #[test]
+    fn test_thread_idx_allocation() {
+        // the thread indexes should be recycled.
+        for _ in 1..=MAX_THREAD_IDX * 2 {
+            std::thread::spawn(|| {
+                set_io_type(IOType::Other);
+            })
+            .join()
+            .unwrap();
+        }
+
+        // use up all available thread index.
+        let pair = Arc::new((Mutex::new(false), Condvar::new()));
+        let mut handles = Vec::new();
+        for _ in 1..=MAX_THREAD_IDX {
+            let pair1 = pair.clone();
+            let h = std::thread::spawn(move || {
+                set_io_type(IOType::Compaction);
+                let (lock, cvar) = &*pair1;
+                let mut stop = lock.lock().unwrap();
+                while !*stop {
+                    stop = cvar.wait(stop).unwrap();
+                }
+                assert_eq!(get_io_type(), IOType::Compaction);
+            });
+            handles.push(h);
+        }
+
+        // the reserved index is used, io type should be IOType::Other
+        for _ in 1..=MAX_THREAD_IDX {
+            std::thread::spawn(|| {
+                set_io_type(IOType::Compaction);
+                assert_eq!(get_io_type(), IOType::Other);
+            })
+            .join()
+            .unwrap();
+        }
+
+        {
+            let (lock, cvar) = &*pair;
+            let mut stop = lock.lock().unwrap();
+            *stop = true;
+            cvar.notify_all();
+        }
+
+        for h in handles {
+            h.join().unwrap();
+        }
+
+        // the thread indexes should be available again.
+        for _ in 1..=MAX_THREAD_IDX {
+            std::thread::spawn(|| {
+                set_io_type(IOType::Compaction);
+                assert_eq!(get_io_type(), IOType::Compaction);
+            })
+            .join()
+            .unwrap();
+        }
+    }
+
+    #[bench]
+    fn bench_write_enable_io_snoop(b: &mut Bencher) {
+        init_io_snooper().unwrap();
+        bench_write(b);
+    }
+
+    #[bench]
+    fn bench_write_disable_io_snoop(b: &mut Bencher) {
+        unsafe { BPF_TABLE = None };
```

Comment:
Can the hook still be attached?

Reply:
the hook is detached when BPF is dropped

Reply:
Then it means we can make `enable-io-snoop` dynamically configurable.

Reply:
good idea, it can be used to debug

Reply:
Some threads call `set_io_type()` only on init, and seems no demand on dynamically config now. So won't support that now.